samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -161,4 +262,140 @@ void handleExpiredRequestWithHostAffinityEnabled(String processorId, String pref
+    LOG.info("Received ControlAction request to move or restart container with processor id {} to host {}",
+        requestMessage.getProcessorId(), requestMessage.getDestinationHost());
+    String processorId = requestMessage.getProcessorId();
+    String destinationHost = requestMessage.getDestinationHost();
+    Pair<ContainerPlacementMessage.StatusCode, String> actionStatus =
+        checkValidControlAction(processorId, destinationHost, requestMessage.getUuid());
+
+    // Request is bad just update the response on message & return
+    if (actionStatus.getKey() == ContainerPlacementMessage.StatusCode.BAD_REQUEST) {
+      return;
+    }
+
+    SamzaResource currentResource = samzaApplicationState.runningProcessors.get(processorId);
+    LOG.info("Processor ID: {} matched a active container with deployment ID: {} running on host: {}", processorId,
+        currentResource.getContainerId(), currentResource.getHost());
+
+    if (destinationHost.equals(ANY_HOST) || !hostAffinityEnabled) {
+      LOG.info("Changing the requested host to {} because either it is requested or host affinity is disabled",
+          ResourceRequestState.ANY_HOST);
+      destinationHost = ANY_HOST;
+    }
+
+    SamzaResourceRequest resourceRequest = containerAllocator.getResourceRequest(processorId, destinationHost);
+    Optional<Duration> requestExpiry =
+        requestMessage.getRequestExpiry() != null ? Optional.of(Duration.ofMillis(requestMessage.getRequestExpiry())) : Optional.empty();
+    ContainerPlacementMetadata actionMetaData =
+        new ContainerPlacementMetadata(requestMessage.getUuid(), processorId, currentResource.getContainerId(),
+            currentResource.getHost(), destinationHost, requestExpiry);
+
+    // Record the resource request for monitoring
+    actionMetaData.setActionStatus(ContainerPlacementMessage.StatusCode.IN_PROGRESS, "Preferred Resources requested");
+    actionMetaData.recordResourceRequest(resourceRequest);
+    actions.put(processorId, actionMetaData);
+    containerAllocator.issueResourceRequest(resourceRequest);
+    LOG.info("Container Placement action with metadata {} and issued a request for resources in progress",
+        actionMetaData);
+  }
+
+  @VisibleForTesting
+  ContainerPlacementMetadata registerContainerPlacementActionForTest(ContainerPlacementRequestMessage requestMessage, ContainerAllocator containerAllocator) {
+    registerContainerPlacementAction(requestMessage, containerAllocator);
+    if (hasActiveContainerPlacementAction(requestMessage.getProcessorId())) {
+      return getPlacementActionMetadata(requestMessage.getProcessorId()).get();
+    }
+    return null;
+  }
+
+  public Optional<Duration> getActionExpiryTimeout(String processorId) {
+    return hasActiveContainerPlacementAction(processorId) ?
+        getPlacementActionMetadata(processorId).get().getRequestActionExpiryTimeout() : Optional.empty();
+  }
```

Comment:
This method is really confusing, it takes as input a processorId then checks if there is action-metadata present on that processorId (which btw is no-longer to be deleted?), then gets the timeout on that timeout. 

Instead maybe the caller can just use getContainerPlacementAction(processorId) to check if there is an inflight action for the SamzaResourceRequest (which is identified by a unique requestId)?